botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1027 @@
+        if not isinstance(input_param, correct_param_type):
+            raise EndpointInputParametersError(
+                msg=f'Input parameter {self.name} is the wrong '
+                f'type. Must be {correct_param_type}'
+            )
+        if self.deprecated is not None:
+            depr_string = f"{self.name} has been deprecated since {self.deprecated['since']}"
+            msg = self.deprecated['message']
+            if msg:
+                depr_string += f"{depr_string}. {msg}"
+            logger.info(depr_string)
+        return None
+
+
+class Ruleset:
+    """A class housing an entire ruleset object. Every ruleset contains a version,
+    parameters (specification of parameters not values) and rules i.e.
+
+    {
+        "version": "1.1",
+        "parameters": {
+            "Region": {
+                ...
+            },
+            "UseDualStack": {
+                ...
+            },
+            "UseFIPS": {
+                ...
+            }
+        },
+        "rules": [
+            {
+                "conditions": [
+                   ...
+                ],
+                "type": "tree",
+                "rules": [
+                    ...
+                ]
+            }
+        ]
+    }
+
+    Additionally, this class is provided input parameters from a client to validate
+    against its defined parameter traits and an instance of an endpoint provider.
+    """
+
+    def __init__(self, version, parameters, rules, input_parameters, provider):
+        self.version = version
+        self.parameters = {
+            name: ParameterDefinition(name, **spec)
+            for name, spec in parameters.items()
+        }
+        self.rules = [
+            RuleCreator.create(
+                assignments=input_parameters, provider=provider, **rule
+            )
+            for rule in rules
+        ]
+        self.input_parameters = input_parameters
+        self.provider = provider
+
+    def validate_input_parameters(self):
+        """Check that each provided value from a client is valid. If not provided,
+        add the default value to the `input_paramters` dictionary.
+        """
+        for param_name, param_spec in self.parameters.items():
+            input_param = self.input_parameters.get(param_name)
+            if input_param is None and param_spec.default is not None:
+                self.input_parameters[param_name] = param_spec.default
+            elif input_param is not None:
+                param_spec.validate_input(input_param)
+        return None
+
+
+class EndpointProvider:
+    """The main interface of this module. Given a rule set and partitions,
+    this class evaluates input parameters against them and returns a resolved
+    endpoint to the client or raises an error.
+    """
+
+    def __init__(self, ruleset_data, partition_data):
+        self.ruleset_data = ruleset_data
+        self.partitions = partition_data
+
+    def get_parameters_definition(self):
```

Comment:
so are you saying this should be cached or attached to the provider as an instance attribute?

Reply:
Yeah, I'd expect it to be computed once for the lifetime of the client. It can be stored here as an attribute probably.

That said, I don't see `get_parameters_definition` used anywhere. Where is this for?

Reply:
This was a request by @jonemo for use in his resolver. Its not a part of the SEP or used anywhere within  the provider

Reply:
The calling code (the future `EndpointResolverv2`) needs to know what parameters to pass into the `resolve()` method. This is where it gets this information from. Could take any other form, such as an instance attribute, just let me know what the final decision is.

Reply:
@jonemo maybe given that all of the parameters are already attached to the provider's rule set object, maybe no API is really necessary at all? Is it still ideal for the resolver to get the parameters as a list instead of a dict?